src/haz3lcore/dynamics/elaborator.re

```diff
@@ -164,6 +164,68 @@ let rec dhexp_of_uexp = (m: Statics.map, uexp: Term.UExp.t): option(DHExp.t) =>
             tl |> List.fold_left((acc, d) => DHExp.Sequence(d, acc), hd)
           };*/
         Some(EmptyHole(id, 0))
+      | Deferral(true, _) => Some(EmptyHole(id, 0))
+      | Deferral(false, _) => Some(InvalidText(0, 0, "_")) // Unbounded deferral
+      | DeferredAp(fn, arg) =>
+        let mk_tuple = (ctor, xs) =>
+          List.length(xs) == 1 ? List.hd(xs) : ctor(xs);
+        let* d_fn = dhexp_of_uexp(m, fn);
+        let ty_fn = Statics.exp_self_typ(m, fn);
+        switch (ty_fn) {
+        | Arrow(ty_arg, ty_ret) =>
+          if (ty_arg == Prod([]) && arg.term == Deferral(true, true)) {
+            Some(d_fn);
+          } else {
+            let args =
+              switch (arg.term) {
+              | Tuple(args) => args
+              | _ => [arg]
+              };
+            let ty_ins =
+              switch (ty_arg) {
+              | Prod(ty_ins) => ty_ins
+              | Unknown(_) as ty_unknown =>
+                List.init(List.length(args), _ => ty_unknown)
+              | _ => [ty_arg]
+              };
+            if (List.length(ty_ins) != List.length(args)) {
+              let+ d_arg = dhexp_of_uexp(m, arg);
+              DHExp.Ap(d_fn, d_arg);
+            } else {
+              // Substitute all deferrals for new variables
+              let+ (pats, d_args, ty_args) =
+                List.combine(args, ty_ins)
+                |> List.fold_left(
```

Comment:
there's no actual data dependency between elements here (i think) so instead of folding, just map, and then use OptUtil.sequence (see other examples in this file)